### DISCUSSION TITLE
type_hints removed from memo in typeguard 4?

### DISCUSSION THREAD

**Discussion by dkamm** [agronholm/typeguard#345] [General]
I noticed that CallMemo in typeguard 3 has a `type_hints` field but TypeCheckMemo in typeguard 4 does not

typeguard 4:
https://github.com/agronholm/typeguard/blob/master/src/typeguard/_memo.py

typeguard 3:
https://github.com/agronholm/typeguard/blob/3.0.2/src/typeguard/_memo.py

Is `type_hints` removal confirmed for typeguard 4?

For context, I'm doing some tensor shape checking and I'd like to include the type hints in my custom error message like so

```
got error:  argument "residual_stack_post" (torch.Tensor) is not an instance of jaxtyping.Float[Tensor, 'b c']
tensor argument info:
        x  shape=(1, 2)  dtype=torch.float32  type_hint=Float[Tensor, 'a b']
        y  shape=(3, 2)  dtype=torch.float32  type_hint=Float[Tensor, 'b c']
        z  shape=(2, 4)  dtype=torch.float32  type_hint=Float[Tensor, 'c d']
```

I'm writing my own `typecheck_fail_callback` and using the memo.locals to construct the error message

**Comment by agronholm** [comment]
Unlike Typeguard 3, Typeguard 4 relies on type hints obtained from static analysis, and directly generates checking code for these hints. For this reason it didn't make sense to store them in the memo. While they could be put back into the memo, this would be costly and only benefit fringe cases such as this.

Is this causing a lot of trouble for you?

**Reply by dkamm** [reply]
Thanks for the clarification!

It's more of a "nice to have" since it would make it easier for users to debug tensor shape mismatches (especially across multiple arguments) at a glance.

Perhaps there could be a mode to enable storing these on memo?  Our use case is research code and I think we'd be ok with the incurring the extra cost.

**Reply by agronholm** [reply]
You could make a feature request, but I am so overburdened with F/OSS work right now that I will likely not work on that request personally. I could accept a PR though.

**Reply by dkamm** [reply]
No worries! I might take a stab at it
